src/main/drivers/adc_stm32f7xx.c

```diff
@@ -83,6 +104,113 @@ const adcTagMap_t adcTagMap[] = {
     { DEFIO_TAG_E__PA7, ADC_DEVICES_12,  ADC_CHANNEL_7  },
 };
 
+void adcInitDevice(adcDevice_t *adcdev, int channelCount)
+{
+    adcdev->ADCHandle.Init.ClockPrescaler        = ADC_CLOCK_SYNC_PCLK_DIV8;
+    adcdev->ADCHandle.Init.ContinuousConvMode    = ENABLE;
+    adcdev->ADCHandle.Init.Resolution            = ADC_RESOLUTION_12B;
+    adcdev->ADCHandle.Init.ExternalTrigConv      = ADC_EXTERNALTRIGCONV_T1_CC1;
+    adcdev->ADCHandle.Init.ExternalTrigConvEdge  = ADC_EXTERNALTRIGCONVEDGE_NONE;
+    adcdev->ADCHandle.Init.DataAlign             = ADC_DATAALIGN_RIGHT;
+    adcdev->ADCHandle.Init.NbrOfConversion       = channelCount;
+#ifdef USE_ADC_INTERNAL
+    // Multiple injected channel seems to require scan conversion mode to be
+    // enabled even if main (non-injected) channel count is 1.
+    adcdev->ADCHandle.Init.ScanConvMode          = ENABLE;
+#else
+    adcdev->ADCHandle.Init.ScanConvMode          = channelCount > 1 ? ENABLE : DISABLE; // 1=scan more that one channel in group
+#endif
+    adcdev->ADCHandle.Init.DiscontinuousConvMode = DISABLE;
+    adcdev->ADCHandle.Init.NbrOfDiscConversion   = 0;
+    adcdev->ADCHandle.Init.DMAContinuousRequests = ENABLE;
+    adcdev->ADCHandle.Init.EOCSelection          = DISABLE;
+    adcdev->ADCHandle.Instance = adcdev->ADCx;
+
+    if (HAL_ADC_Init(&adcdev->ADCHandle) != HAL_OK)
+    {
+      /* Initialization Error */
+    }
+}
+
+static adcDevice_t adc;
+
+#ifdef USE_ADC_INTERNAL
+
+static adcDevice_t adcInternal;
+static ADC_HandleTypeDef *adcInternalHandle;
+
+void adcInitInternalInjected(adcDevice_t *adcdev)
+{
+    adcInternalHandle = &adcdev->ADCHandle;
+
+    ADC_InjectionConfTypeDef iConfig;
+
+    iConfig.InjectedChannel      = ADC_CHANNEL_VREFINT;
+    iConfig.InjectedRank         = 1;
+    iConfig.InjectedSamplingTime = ADC_SAMPLETIME_480CYCLES;
+    iConfig.InjectedOffset       = 0;
+    iConfig.InjectedNbrOfConversion = 2;
+    iConfig.InjectedDiscontinuousConvMode = DISABLE;
+    iConfig.AutoInjectedConv     = DISABLE;
+    iConfig.ExternalTrigInjecConv = 0;     // Don't care
+    iConfig.ExternalTrigInjecConvEdge = 0; // Don't care
+
+    if (HAL_ADCEx_InjectedConfigChannel(adcInternalHandle, &iConfig) != HAL_OK) {
+        /* Channel Configuration Error */
+    }
+
+    iConfig.InjectedChannel      = ADC_CHANNEL_TEMPSENSOR;
+    iConfig.InjectedRank         = 2;
+
+    if (HAL_ADCEx_InjectedConfigChannel(adcInternalHandle, &iConfig) != HAL_OK) {
+        /* Channel Configuration Error */
+    }
+
+    adcVREFINTCAL = *(uint16_t *)VREFINT_CAL_ADDR;
```

Comment:
It's already typed

Reply:
- Definitions of calibration data addresses in LL driver are incorrect.